### PR TITLE
Issue #10809: RANGE Hint Corrections

### LINKED ISSUE #10809: Window function results should not be affected by row order when using INTERVAL RANGE with dates
### What happens?

When using a window function with range such as `RANGE BETWEEN INTERVAL 3 DAYS PRECEDING AND INTERVAL 1 DAYS`, duckdb takes the order of rows into account instead of purely the date range specified, and produces different result from the windows function based on the row order. 

This shouldn't happen because the aggregate function in window should be applied on all the rows with dates in range, rather than being affected by the row order. it seems like window function is only applied to preceding rows that satisfy the interval constraint.

This issue only starts appearing from duckdb version `0.9.0` and above. 

### To Reproduce

given table:
| id | requested_date | city | wait_time |
|-----------|--------------------------|------------------------|---------------------|
| 0         | 2023-01-05               | San Francisco       | 2925            |
| 1         | 2023-01-03               | San Francisco       | 755             |
| 2         | 2023-01-03               | San Francisco       | 2880            |
| 3         | 2023-01-05               | San Francisco       | 1502            |
| 4         | 2023-01-03               | San Francisco       | 2900            |
| 5         | 2023-01-01               | San Francisco       | 1210            |
| 6         | 2023-01-04               | San Francisco       | 200             |
| 7         | 2023-01-02               | San Francisco       | 980             |
| 8         | 2023-01-02               | San Francisco       | 430             |
| 9         | 2023-01-05               | San Francisco       | 2999            |
| 10        | 2023-01-01               | San Francisco       | 856             |
| 11        | 2023-01-02               | San Francisco       | 490             |
| 12        | 2023-01-02              | San Francisco       | 720             |

If we randomize the row order in a table using 
`SELECT * FROM TABLE ORDER BY random()`

Then running this query on the below table will give different results
`SELECT "requested_date",
       "city",
       (min("wait_time"::DOUBLE) OVER win_3d)::DOUBLE
  FROM table WINDOW win_3d AS (PARTITION BY "city" ORDER BY requested_date ASC RANGE BETWEEN INTERVAL 3 DAYS PRECEDING AND INTERVAL 1 DAYS PRECEDING)
 ORDER BY "requested_date",
          "city"
`

### OS:

OS 14.1, Apple M3 Pro

### DuckDB Version:

0.10.0

### DuckDB Client:

Python

### Full Name:

Melody Lui

### Affiliation:

Delphina

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by hawkfish**
Correct the rules for the RANGE search optimisation to handle empty prior ranges
and fix a fencepost error on range ends.

fixes: #10809
fixes: duckdblabs/duckdb-internal#1345

### CODE REVIEW

File: src/execution/window_executor.cpp [hunk 1]
```diff
@@ -207,19 +207,24 @@ static idx_t FindTypedRangeBound(const WindowInputColumn &over, const idx_t orde
...
 
-	if (order_begin < prev.start && prev.start < order_end) {
-		const auto first = over.GetCell<T>(prev.start);
-		if (!comp(val, first)) {
-			//	prev.first <= val, so we can start further forward
-			begin += (prev.start - order_begin);
+	//	Try to reuse the previous bounds to restrict the search.
+	//	This is only valid if the previous bounds were non-empty
+	//	Only inject the comparisons if the previous bounds are a strict subset.
+	if (prev.start < prev.end) {
```
Comment (taniabogatsch) [inline src/execution/window_executor.cpp:213]: Why is this condition false, if we randomize the row order? And we could have an early-out in that case.

File: src/execution/window_executor.cpp [hunk 2]
```diff
@@ -207,19 +207,24 @@ static idx_t FindTypedRangeBound(const WindowInputColumn &over, const idx_t orde
...
-	if (order_begin <= prev.end && prev.end < order_end) {
-		const auto second = over.GetCell<T>(prev.end);
-		if (!comp(second, val)) {
-			//	val <= prev.second, so we can end further back
-			// (prev.second is the largest peer)
-			end -= (order_end - prev.end - 1);
+		if (order_begin < prev.end && prev.end < order_end) {
+			const auto second = over.GetCell<T>(prev.end - 1);
+			if (!comp(second, val)) {
+				//	val <= prev.second, so we can end further back
```
Comment (taniabogatsch) [inline src/execution/window_executor.cpp:224]: This is now strictly less than `prev.second`?

File: src/execution/window_executor.cpp (hunk 1 above)
Comment (hawkfish) [reply src/execution/window_executor.cpp:213]: The randomisation just changes which peer row comes first and that can change the result - even making it correct if the min row happens to come first.

File: src/execution/window_executor.cpp (hunk 2 above)
Comment (hawkfish) [reply src/execution/window_executor.cpp:224]: Yes - it wasn't before.

File: src/execution/window_executor.cpp (hunk 1 above)
Comment (hawkfish) [reply src/execution/window_executor.cpp:213]: > And we could have an early-out in that case 

This is a performance optimisation - the test is for the _previous_ range points, not the current ones. We always have to check the current ones.

Comment (taniabogatsch) [reply src/execution/window_executor.cpp:213]: Okay, thanks! What tipped me off is that we're comparing `start < end` of the same `prev`.

**Review by taniabogatsch — APPROVED** [review]

**Comment by taniabogatsch** [comment]
@ hawkfish, LGTM. Let's un-draft it.

**Pull request merged**